lib/dynflow/execution_plan.rb

```diff
@@ -505,6 +509,10 @@ def self.steps_from_hash(step_ids, execution_plan_id, world)
       end
     end
 
+    def unlock_all_singleton_locks!
+      actions.select(&:holds_singleton_lock?).each(&:singleton_unlock!)
```

Comment:
Loading all actions every time the plan goes to paused/stopped state might be quite resource-hungry. Going directly though coordinator API to just release the singleton locks by `execution_plan-#{id}` owner would be probably much faster.